source/adapters/level_zero/context.hpp

```diff
@@ -279,24 +283,28 @@ struct ur_context_handle_t_ : _ur_object {
         auto EventCachesMap =
             WithProfiling ? &EventCachesDeviceMap[0] : &EventCachesDeviceMap[1];
         if (EventCachesMap->find(Device) == EventCachesMap->end()) {
-          EventCaches.emplace_back();
-          (*EventCachesMap)[Device] = &EventCaches.back();
+          std::list<ur_event_handle_t> *deviceEventCache =
+              new std::list<ur_event_handle_t>;
+          EventCaches.push_back(deviceEventCache);
+          (*EventCachesMap)[Device] = deviceEventCache;
         }
         return (*EventCachesMap)[Device];
       } else {
-        return WithProfiling ? &EventCaches[0] : &EventCaches[1];
+        return WithProfiling ? EventCaches[0] : EventCaches[1];
```

Comment:
This line is causing a crash in the sycl e2e tests [here](Plugin/interop-level-zero.cpp
). I've run a RelWithDebInfo build locally and got this output from `gdb` which appears to show an invalid access of the `EventCaches` vector. Does this code need to handle the empty case?

```
(gdb) run
Starting program: /home/benie/Projects/intel/llvm/interop-level-zero.cpp.tmp.out
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff2fff640 (LWP 171815)]
[New Thread 0x7ffff11ff640 (LWP 171816)]
[New Thread 0x7ffff09fe640 (LWP 171817)]
[Thread 0x7ffff09fe640 (LWP 171817) exited]
/usr/include/c++/11/bits/stl_vector.h:1045: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = std::__cxx11::list<ur_event_handle_t_*>*; _Alloc = std::allocator<std::__cxx11::list<ur_event_handle_t_*>*>; std::vector<_Tp, _Alloc>::reference = std::__cxx11::list<ur_event_handle_t_*>*&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__n < this->size()' failed.

Thread 1 "interop-level-z" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737349939840) at ./nptl/pthread_kill.c:44
44      ./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737349939840) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737349939840) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737349939840, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff4242476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff42287f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff7b0a368 in std::__replacement_assert (__file=__file@entry=0x7ffff7b7ed30 "/usr/include/c++/11/bits/stl_vector.h",
    __line=__line@entry=1045,
    __function=__function@entry=0x7ffff7b824a8 "std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = std::__cxx11::list<ur_event_handle_t_*>*; _Alloc = std::allocator<std::__cxx11"...,
    __condition=__condition@entry=0x7ffff7b7e194 "__n < this->size()") at /usr/include/x86_64-linux-gnu/c++/11/bits/c++config.h:514
#6  0x00007ffff7b1578e in std::vector<std::__cxx11::list<ur_event_handle_t_*, std::allocator<ur_event_handle_t_*> >*, std::allocator<std::__cxx11::list<ur_event_handle_t_*, std::allocator<ur_event_handle_t_*> >*> >::operator[] (this=0x24e3990, __n=1)
    at /usr/include/c++/11/bits/stl_vector.h:1045
#7  std::vector<std::__cxx11::list<ur_event_handle_t_*, std::allocator<ur_event_handle_t_*> >*, std::allocator<std::__cxx11::list<ur_event_handle_t_*, std::allocator<ur_event_handle_t_*> >*> >::operator[] (__n=1, this=0x24e3990) at /usr/include/c++/11/bits/stl_vector.h:1043
#8  ur_context_handle_t_::getEventCache[abi:cxx11](bool, bool, ur_device_handle_t_*) (this=this@entry=0x24e3570,
    HostVisible=HostVisible@entry=true, WithProfiling=WithProfiling@entry=false, Device=<optimised out>)
    at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/context.hpp:293
#9  0x00007ffff7b15839 in ur_context_handle_t_::getEventFromContextCache (this=0x24e3570, HostVisible=<optimised out>,
    WithProfiling=<optimised out>, Device=<optimised out>)
    at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/context.cpp:553
#10 0x00007ffff7b22e32 in EventCreate (Context=0x24e3570, Queue=<optimised out>, IsMultiDevice=<optimised out>,
    HostVisible=<optimised out>, RetEvent=0x7fffffffc8d0)
    at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/event.cpp:1062
#11 0x00007ffff7b4f737 in createEventAndAssociateQueue (Queue=0x24e43d0, Event=0x7fffffffc8d0, CommandType=2147483646, CommandList=...,
    IsInternal=<optimised out>, IsMultiDevice=<optimised out>, HostVisible=std::optional<bool> [no contained value])
    at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/queue.cpp:1519
#12 0x00007ffff7b5183a in ur_queue_handle_t_::executeCommandList (this=0x24e43d0, CommandList=..., IsBlocking=<optimised out>,
    OKToBatchCommand=<optimised out>) at /usr/include/c++/11/bits/move.h:77
#13 0x00007ffff7b31cd0 in enqueueMemCopyHelper (CommandType=UR_COMMAND_MEM_BUFFER_WRITE, Queue=0x24e43d0, Dst=0xffffd556aa1f0000,
    BlockingWrite=<optimised out>, Size=4, Src=0x7fffffffe07c, NumEventsInWaitList=0, EventWaitList=0x0, OutEvent=0x1a71bb8,
    PreferCopyEngine=true) at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/memory.cpp:86
#14 0x00007ffff7b3501c in urEnqueueMemBufferWrite (Queue=0x24e43d0, hBuffer=<optimised out>, blockingWrite=<optimised out>, offset=0,
    size=4, pSrc=0x7fffffffe07c, numEventsInWaitList=0, phEventWaitList=0x0, phEvent=0x1a71bb8)
    at /home/benie/Projects/intel/llvm/build-rd/_deps/unified-runtime-src/source/adapters/level_zero/memory.cpp:530
#15 0x00007ffff7b59076 in pi2ur::piEnqueueMemBufferWrite (OutEvent=0x1a71bb8, EventsWaitList=0x0, NumEventsInWaitList=0,
    Ptr=<optimised out>, Size=<optimised out>, Offset=<optimised out>, BlockingWrite=<optimised out>, Buffer=<optimised out>,
    Queue=<optimised out>) at /home/benie/Projects/intel/llvm/sycl/plugins/level_zero/../unified_runtime/pi2ur.hpp:3959
#16 piEnqueueMemBufferWrite (Queue=<optimised out>, Buffer=<optimised out>, BlockingWrite=<optimised out>, Offset=<optimised out>,
    Size=<optimised out>, Ptr=<optimised out>, NumEventsInWaitList=0, EventWaitList=0x0, Event=0x1a71bb8)
    at /home/benie/Projects/intel/llvm/sycl/plugins/level_zero/pi_level_zero.cpp:715
#17 0x00007ffff498bcad in sycl::_V1::detail::plugin::call_nocheck<(sycl::_V1::detail::PiApiKind)81, _pi_queue*, _pi_mem*, unsigned int, unsigned long, unsigned long, char*, unsigned long, _pi_event**, _pi_event**> (this=this@entry=0x496910)
    at /home/benie/Projects/intel/llvm/build-rd/include/sycl/detail/pi.def:112
#18 0x00007ffff498dc6c in sycl::_V1::detail::plugin::call<(sycl::_V1::detail::PiApiKind)81, _pi_queue*, _pi_mem*, unsigned int, unsigned long, unsigned long, char*, unsigned long, _pi_event**, _pi_event**> (this=0x496910)
    at /home/benie/Projects/intel/llvm/sycl/source/detail/plugin.hpp:257
#19 sycl::_V1::detail::copyH2D (SYCLMemObj=SYCLMemObj@entry=0x24dd520, SrcMem=SrcMem@entry=0x7fffffffe07c "\002", DimSrc=DimSrc@entry=1,
    SrcSize=..., SrcAccessRange=..., SrcOffset=..., SrcElemSize=SrcElemSize@entry=4, DstMem=DstMem@entry=0x24ccd90,
    TgtQueue=std::shared_ptr<sycl::_V1::detail::queue_impl> (use count 12, weak count 12) = {...}, DimDst=1, DstSize=..., DstAccessRange=...,
    DstOffset=..., DstElemSize=4, DepEvents=std::vector of length 0, capacity 0, OutEvent=@0x1a71bb8: 0x1a67ea0,
    OutEventImpl=std::shared_ptr<sycl::_V1::detail::event_impl> (use count 2, weak count 0) = {...})
    at /home/benie/Projects/intel/llvm/sycl/source/detail/memory_manager.cpp:524
#20 0x00007ffff4995d2b in sycl::_V1::detail::MemoryManager::copy (SYCLMemObj=0x24dd520, SrcMem=0x7fffffffe07c, SrcQueue=..., DimSrc=1,
    SrcSize=..., SrcAccessRange=..., SrcOffset=..., SrcElemSize=4, DstMem=0x24ccd90,
    TgtQueue=std::shared_ptr<sycl::_V1::detail::queue_impl> (empty) = {...}, DimDst=1, DstSize=..., DstAccessRange=..., DstOffset=...,
    DstElemSize=4, DepEvents=std::vector of length 0, capacity 0, OutEvent=@0x1a71bb8: 0x1a67ea0,
    OutEventImpl=std::shared_ptr<sycl::_V1::detail::event_impl> (use count 2, weak count 0) = {...})
    at /home/benie/Projects/intel/llvm/sycl/source/detail/memory_manager.cpp:780
#21 0x00007ffff4a0fa7b in sycl::_V1::detail::MemCpyCommand::enqueueImp (this=0x24ed6a0)
    at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/commands.cpp:1530
#22 0x00007ffff4a1b255 in sycl::_V1::detail::Command::enqueue (this=0x24ed6a0, EnqueueResult=..., Blocking=<optimised out>,
    ToCleanUp=...) at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/commands.cpp:875
#23 0x00007ffff4a322c3 in sycl::_V1::detail::Scheduler::GraphProcessor::enqueueCommand (Cmd=0x24ed6a0, GraphReadLock=...,
    EnqueueResult=..., ToCleanUp=std::vector of length 0, capacity 0, RootCommand=0x1aa2ac0, Blocking=sycl::_V1::detail::NON_BLOCKING)
    at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/graph_processor.cpp:144
#24 0x00007ffff4a3229a in sycl::_V1::detail::Scheduler::GraphProcessor::enqueueCommand (Cmd=0x1aa2ac0, GraphReadLock=...,
    EnqueueResult=..., ToCleanUp=std::vector of length 0, capacity 0, RootCommand=0x1aa2ac0, Blocking=sycl::_V1::detail::NON_BLOCKING)
    at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/graph_processor.cpp:126
#25 0x00007ffff4a2f3cb in sycl::_V1::detail::Scheduler::enqueueCommandForCG (this=0x192b120, NewEvent=..., AuxiliaryCmds=...,
    Blocking=sycl::_V1::detail::NON_BLOCKING) at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/scheduler.cpp:206
#26 0x00007ffff4a315ae in sycl::_V1::detail::Scheduler::addCG (this=0x192b120,
    CommandGroup=std::unique_ptr<sycl::_V1::detail::CG> = {...},
    Queue=std::shared_ptr<sycl::_V1::detail::queue_impl> (use count 12, weak count 12) = {...}, CommandBuffer=CommandBuffer@entry=0x0,
    Dependencies=std::vector of length 0, capacity 0) at /home/benie/Projects/intel/llvm/sycl/source/detail/scheduler/scheduler.cpp:152
#27 0x00007ffff4a73dcf in sycl::_V1::handler::finalize (this=<optimised out>)
    at /home/benie/Projects/intel/llvm/sycl/source/handler.cpp:554
#28 0x00007ffff49e80da in sycl::_V1::detail::queue_impl::finalizeHandler<sycl::_V1::handler> (this=<optimised out>, Handler=...,
    EventRet=...) at /home/benie/Projects/intel/llvm/sycl/source/detail/queue_impl.hpp:786
#29 0x00007ffff49ed3de in sycl::_V1::detail::queue_impl::submit_impl(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*) (this=this@entry=0x22a16c0,
    CGF=..., Self=..., PrimaryQueue=..., SecondaryQueue=std::shared_ptr<sycl::_V1::detail::queue_impl> (empty) = {...}, Loc=...,
    PostProcess=0x0) at /home/benie/Projects/intel/llvm/sycl/source/detail/queue_impl.hpp:848
#30 0x00007ffff4a94e96 in sycl::_V1::detail::queue_impl::submit(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*) (
    PostProcess=0x0, Loc=..., Self=..., CGF=..., this=0x22a16c0) at /home/benie/Projects/intel/llvm/sycl/source/detail/queue_impl.hpp:430
#31 sycl::_V1::queue::submit_impl(std::function<void (sycl::_V1::handler&)>, sycl::_V1::detail::code_location const&) (
    this=<optimised out>, CGH=..., CodeLoc=...) at /home/benie/Projects/intel/llvm/sycl/source/queue.cpp:180
#32 0x0000000000405311 in sycl::_V1::queue::submit<main::{lambda(sycl::_V1::handler&)#1}>(main::{lambda(sycl::_V1::handler&)#1}, sycl::_V1::detail::code_location const&) (this=0x7fffffffe0d0, CGF=..., CodeLoc=...) at build-rd/bin/../include/sycl/queue.hpp:366
#33 0x000000000040508d in main () at sycl/test-e2e/Plugin/interop-level-zero.cpp:76
```

Reply:
Thanks @kbenzie ,

I will take a look. This structure used to be a static variable, but now it cannot be. If the Context init is not called, then this will not be init, I may need to move the init of this to another object if context is not always init.

Reply:
perhaps something along these lines is necessary but at a different interception point:

* https://github.com/intel/llvm/pull/12432
* https://github.com/intel/llvm/pull/12619
* https://github.com/intel/llvm/pull/12515